poly/src/polynomial/mod.rs

```diff
@@ -1,164 +1,93 @@
-//! Work with sparse and dense polynomials.
-
-use crate::{EvaluationDomain, Evaluations};
-use ark_ff::{FftField, Field};
-use ark_std::{borrow::Cow, convert::TryInto, vec::Vec};
-use DenseOrSparsePolynomial::*;
-
-mod dense;
-mod sparse;
-
-pub use dense::DensePolynomial;
-pub use sparse::SparsePolynomial;
-
-/// Represents either a sparse polynomial or a dense one.
-#[derive(Clone)]
-pub enum DenseOrSparsePolynomial<'a, F: Field> {
-    /// Represents the case where `self` is a sparse polynomial
-    SPolynomial(Cow<'a, SparsePolynomial<F>>),
-    /// Represents the case where `self` is a dense polynomial
-    DPolynomial(Cow<'a, DensePolynomial<F>>),
-}
+//! Modules for working with univariate or multivariate polynomials.
+
+use ark_ff::Field;
+use ark_std::{
+    fmt::Debug,
+    hash::Hash,
+    ops::{AddAssign, Index, SubAssign},
+    vec::Vec,
+};
+use rand::Rng;
+
+pub mod multivariate;
+pub mod univariate;
+
+/// Describes the common interface for univariate and multivariate polynomials
+pub trait Polynomial<F: Field>:
+    Sized
+    + Clone
+    + Debug
+    + Default
+    + Hash
+    + PartialEq
+    + Eq
+    + Send
+    + Sync
+    + for<'a> AddAssign<&'a Self>
+    + for<'a> AddAssign<(F, &'a Self)>
+    + for<'a> SubAssign<&'a Self>
+{
+    /// The domain of the polynomial.
+    type Domain: Sized + Clone + Ord + Debug + Sync;
+
+    /// Returns the zero polynomial.
+    fn zero() -> Self;
 
-impl<'a, F: 'a + Field> From<DensePolynomial<F>> for DenseOrSparsePolynomial<'a, F> {
-    fn from(other: DensePolynomial<F>) -> Self {
-        DPolynomial(Cow::Owned(other))
-    }
-}
+    /// Checks if the given polynomial is zero.
+    fn is_zero(&self) -> bool;
 
-impl<'a, F: 'a + Field> From<&'a DensePolynomial<F>> for DenseOrSparsePolynomial<'a, F> {
-    fn from(other: &'a DensePolynomial<F>) -> Self {
-        DPolynomial(Cow::Borrowed(other))
-    }
-}
+    /// Returns the total degree of the polynomial
+    fn degree(&self) -> usize;
 
-impl<'a, F: 'a + Field> From<SparsePolynomial<F>> for DenseOrSparsePolynomial<'a, F> {
-    fn from(other: SparsePolynomial<F>) -> Self {
-        SPolynomial(Cow::Owned(other))
-    }
-}
+    /// Evaluates `self` at the given `point` in `Self::Domain`.
+    fn evaluate(&self, point: &Self::Domain) -> F;
 
-impl<'a, F: Field> From<&'a SparsePolynomial<F>> for DenseOrSparsePolynomial<'a, F> {
-    fn from(other: &'a SparsePolynomial<F>) -> Self {
-        SPolynomial(Cow::Borrowed(other))
-    }
-}
+    /// If `num_vars` is `None`, outputs a polynomial a univariate polynomial
+    /// of degree `d` where each coefficient is sampled uniformly at random.
+    ///
+    /// If `num_vars` is `Some(l)`, outputs an `l`-variate polynomial which
+    /// is the sum of `l` `d`-degree univariate polynomials where each coefficient
+    /// is sampled uniformly at random.
+    fn rand<R: Rng>(d: usize, num_vars: Option<usize>, rng: &mut R) -> Self;
```

Comment:
I presume that `num_vars` is required for hiding in MarlinPST13? Can we move it to a helper method outside the trait?